spark/sql-13/src/main/scala/org/elasticsearch/spark/sql/DataFrameValueWriter.scala

```diff
@@ -157,6 +170,80 @@ class DataFrameValueWriter(writeUnknownTypes: Boolean = false) extends Filtering
     Result.SUCCESFUL()
   }
 
+  private def writeMapWithInferredSchema(value: Any, generator: Generator): Result = {
+    value match {
+      case sm: SMap[_, _] => doWriteMapWithInferredSchema(sm, generator)
+      case jm: JMap[_, _] => doWriteMapWithInferredSchema(jm.asScala, generator)
+      // unknown map type
+      case _              => handleUnknown(value, generator)
+    }
+  }
+
+  private def doWriteMapWithInferredSchema(map: SMap[_, _], generator: Generator): Result = {
+    if (map != null && map.valuesIterator.hasNext) {
+      val sampleValueOption = getFirstNotNullElement(map.valuesIterator)
+      val schema = inferMapSchema(sampleValueOption)
+      doWriteMap(schema, map, generator)
+    } else {
+      writeEmptyMap(generator)
+    }
+  }
+
+  private def writeEmptyMap(generator: Generator): Result = {
+    generator.writeBeginObject().writeEndObject()
+    Result.SUCCESFUL()
+  }
+
+  private def inferMapSchema(valueOption: Option[Any]): MapType = {
+    if(valueOption.isDefined) {
+      val valueType = inferType(valueOption.get)
+      MapType(StringType, valueType) //The key type is never read
+    } else {
+      MapType(StringType, StringType) //Does not matter if the map is empty or has no values
+    }
+  }
+
+  def inferArraySchema(array: Array[_]): DataType = {
+    val EMPTY_ARRAY_TYPE = StringType  //Makes no difference for an empty array
+    if (array.isEmpty) {
+      EMPTY_ARRAY_TYPE
+    } else {
+      val sampleValueOption = getFirstNotNullElement(array.iterator)
+      if (sampleValueOption.isDefined) {
+        inferType(sampleValueOption.get)
+      }
+      else {
+        EMPTY_ARRAY_TYPE
+      }
+    }
+  }
+
+  def getFirstNotNullElement(iterator: Iterator[_]): Option[Any] = {
+    iterator.find(value => Option(value).isDefined)
+  }
+
+  private def inferType(value: Any): DataType = {
+    value match {
+      case _: String            => StringType
+      case Int                  => IntegerType
```

Comment:
Scala question: Should this be `_: Int` ?

Reply:
And I'm pretty sure the java ones are redundant, but I'll leave them in because they don't hurt anything.